File: src/imagery/i.hyper/libsixsv/examples/05_spatial_filters.c
/**
 * 05_spatial_filters.c — Spatial filtering, terrain, and adjacency corrections.
 *
 * Demonstrates:
 *  - Gaussian and box spatial filters (NaN-safe, OpenMP + GPU-offload)
 *  - Terrain illumination and transmittance correction (Proy 1989)
 *  - Adjacency effect correction (Vermote 1997)
 *  - Uncertainty propagation (noise + AOD perturbation in quadrature)
 *
 * Compile:
 *   gcc -std=c11 -O2 -fopenmp -I../include 05_spatial_filters.c \
 *       -L../testsuite -lsixsv -lm -fopenmp -Wl,-rpath,../testsuite -o 05_spatial_filters
 */

#include <stdio.h>
#include <stdlib.h>
#define _USE_MATH_DEFINES
#include <math.h>
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif
#include "atcorr.h"
#include "spatial.h"
#include "terrain.h"
#include "adjacency.h"
#include "uncertainty.h"

int main(void)
{
    const int nrows = 64, ncols = 64, npix = nrows * ncols;

    /* ── Synthetic AOD map with a few NaN outliers ───────────────────────── */
    float *aod_map = malloc(npix * sizeof(float));
    for (int r = 0; r < nrows; r++)
        for (int c = 0; c < ncols; c++) {
            int i = r * ncols + c;
            aod_map[i] = 0.10f + 0.15f * sinf(r * 0.3f) * cosf(c * 0.3f);
        }
    /* Inject NaN outliers (e.g. cloud-masked pixels) */
    for (int i = 10; i < npix; i += 100)
        aod_map[i] = (float)NAN;   /* cloud-masked outlier */

    /* ── 1. Gaussian filter (in-place, NaN excluded from neighbourhood) ──── */
    float *aod_gauss = malloc(npix * sizeof(float));
    for (int i = 0; i < npix; i++) aod_gauss[i] = aod_map[i];
    spatial_gaussian_filter(aod_gauss, nrows, ncols, /*sigma=*/2.5f);
    /* Count NaN pixels that were healed */
    int n_nan_before = 0, n_nan_after = 0;
    for (int i = 0; i < npix; i++) {
        if (aod_map[i] != aod_map[i]) n_nan_before++;
        if (aod_gauss[i] != aod_gauss[i]) n_nan_after++;
    }
    printf("Gaussian filter: NaN before=%d  after=%d\n",
           n_nan_before, n_nan_after);

    /* ── 2. Box filter (separable, allocates separate output buffer) ─────── */
    float *aod_box = malloc(npix * sizeof(float));
    spatial_box_filter(aod_map, aod_box, nrows, ncols, /*filter_half=*/3);
    float box_mean = 0.0f; int box_n = 0;
    for (int i = 0; i < npix; i++)
        if (aod_box[i] == aod_box[i]) { box_mean += aod_box[i]; box_n++; }
    printf("Box filter (half=3): output mean=%.4f\n",
           box_n ? box_mean / box_n : 0.0f);

    /* ── 3. Terrain illumination correction ──────────────────────────────── */
    float sza_deg = 35.0f, saa_deg = 150.0f;   /* solar position */
    float slope_deg = 20.0f, aspect_deg = 180.0f;  /* south-facing slope */

    float cos_i = cos_incidence(sza_deg, saa_deg, slope_deg, aspect_deg);
    float V_d   = skyview_factor(slope_deg);
    printf("\nTerrain:  cos_i = %.4f  skyview = %.4f  %s\n",
           cos_i, V_d, cos_i <= 0.0f ? "(topographic shadow)" : "");

    /* Adjust T_down for the tilted surface (example values from LUT) */
    float T_down     = 0.85f;  /* total downward transmittance */
    float T_down_dir = 0.70f;  /* direct component */
    float cos_sza    = cosf(sza_deg * M_PI / 180.0f);

    float T_down_eff = atcorr_terrain_T_down(T_down, T_down_dir, cos_sza, cos_i, V_d);
    printf("T_down flat=%.4f  T_down_eff (slope)=%.4f\n", T_down, T_down_eff);

    /* Per-pixel T_up correction for view-zenith variation (tilted surface) */
    float T_up       = 0.90f;
    float cos_vza    = cosf(8.0f * M_PI / 180.0f);
    float T_up_eff   = atcorr_terrain_T_up(T_up, cos_vza, /*vza_pixel_deg=*/15.0f);
    printf("T_up flat=%.4f   T_up_eff (off-nadir view)=%.4f\n", T_up, T_up_eff);

    /* ── 4. Adjacency effect correction ──────────────────────────────────── */
    /* Simulate BOA reflectance map for one band */
    float *r_boa = malloc(npix * sizeof(float));
    for (int i = 0; i < npix; i++)
        r_boa[i] = 0.10f + 0.25f * (float)(i % ncols) / ncols;

    /* Apply Vermote 1997 in-place (computes r_env and T_dir internally) */
    adjacency_correct_band(
        r_boa, nrows, ncols,
        /*psf_radius_km=*/1.0f,       /* adjacency PSF radius */
        /*pixel_size_m=*/30.0f,       /* Landsat 30 m pixels */
        /*T_scat=*/1.0f - T_down_dir, /* scattering transmittance */
        /*s_alb=*/0.10f,
        /*wl_um=*/0.65f,
        /*aod550=*/0.15f,
        /*pressure=*/1013.25f,
        /*sza_deg=*/sza_deg,
        /*vza_deg=*/8.0f);

    float r_mean = 0.0f;
    for (int i = 0; i < npix; i++) r_mean += r_boa[i];
    printf("\nAdjacency correction: mean BOA = %.4f\n", r_mean / npix);

    /* ── 5. Uncertainty propagation ──────────────────────────────────────── */
    float wl_um  = 0.65f;
    float aod_val = 0.15f, h2o_val = 2.0f;
    float d2 = (float)sixs_earth_sun_dist2(185);
    float E0 = sixs_E0(wl_um);

    /* Build a minimal LUT for the uncertainty call */
    float wl_u[]  = {0.65f};
    float aod_u[] = {0.0f, 0.1f, 0.2f, 0.4f};
    float h2o_u[] = {1.0f, 3.0f};
    int Nu = 4 * 2 * 1;
    float *Ra = calloc(Nu, sizeof(float));
    float *Td = calloc(Nu, sizeof(float));
    float *Tu = calloc(Nu, sizeof(float));
    float *sa = calloc(Nu, sizeof(float));
    LutConfig cfg_u = {
        .wl=wl_u, .n_wl=1, .aod=aod_u, .n_aod=4, .h2o=h2o_u, .n_h2o=2,
        .sza=35.0f, .vza=8.0f, .raa=90.0f, .altitude_km   = 1000.0f,
        .atmo_model=ATMO_US62, .aerosol_model=AEROSOL_CONTINENTAL, .ozone_du=300.0f,
    };
    LutArrays lut_u = { .R_atm=Ra, .T_down=Td, .T_up=Tu, .s_alb=sa };
    atcorr_compute_lut(&cfg_u, &lut_u);

    float *sigma_out = malloc(npix * sizeof(float));
    uncertainty_compute_band(
        /*rad_band=*/NULL, r_boa, npix,
        E0, d2, cosf(35.0f * M_PI / 180.0f),
        /*T_down=*/0.85f, /*T_up=*/0.90f, /*s_alb=*/0.08f, /*R_atm=*/0.05f,
        /*nedl=*/0.0f,   /* estimate from rad_band — NULL → use 0.01 fallback */
        /*aod_sigma=*/0.04f,
        &cfg_u, &lut_u, wl_um, aod_val, h2o_val,
        sigma_out);

    float sig_mean = 0.0f;
    for (int i = 0; i < npix; i++) sig_mean += sigma_out[i];
    printf("Uncertainty (σ_rfl) at 650 nm: mean = %.5f\n", sig_mean / npix);

    free(aod_map); free(aod_gauss); free(aod_box);
    free(r_boa); free(sigma_out);
    free(Ra); free(Td); free(Tu); free(sa);
    return 0;
}

File: src/imagery/i.hyper/libsixsv/examples/05_spatial_filters.py
"""
05_spatial_filters.py — Spatial filtering, terrain, and adjacency corrections.

Demonstrates:
  - Gaussian and box spatial filters (NaN-safe)
  - Terrain illumination and transmittance corrections
  - Adjacency effect correction (Vermote 1997)
  - Uncertainty propagation (noise + AOD perturbation in quadrature)

Run:
    python3 05_spatial_filters.py
"""

import os, ctypes, math
import numpy as np

# ── Load libsixsv.so ──────────────────────────────────────────────────────────
_here = os.path.dirname(os.path.abspath(__file__))
for _p in [os.environ.get("LIB_SIXSV", ""),
           os.path.join(_here, "..", "testsuite", "libsixsv.so")]:
    if _p and os.path.exists(_p):
        _lib = ctypes.CDLL(_p); break
else:
    raise ImportError("libsixsv.so not found — run 'make' in testsuite/ first.")

_FP = ctypes.POINTER(ctypes.c_float)

class _Cfg(ctypes.Structure):
    _fields_ = [
        ("wl",  _FP), ("n_wl",  ctypes.c_int),
        ("aod", _FP), ("n_aod", ctypes.c_int),
        ("h2o", _FP), ("n_h2o", ctypes.c_int),
        ("sza", ctypes.c_float), ("vza", ctypes.c_float),
        ("raa", ctypes.c_float), ("altitude_km", ctypes.c_float),
        ("atmo_model",    ctypes.c_int), ("aerosol_model",    ctypes.c_int),
        ("surface_pressure", ctypes.c_float), ("ozone_du", ctypes.c_float),
        ("mie_r_mode",   ctypes.c_float), ("mie_sigma_g", ctypes.c_float),
        ("mie_m_real",   ctypes.c_float), ("mie_m_imag",  ctypes.c_float),
        ("brdf_type",    ctypes.c_int),   ("brdf_params",  ctypes.c_float * 5),
        ("enable_polar", ctypes.c_int),
    ]

class _Arr(ctypes.Structure):
    _fields_ = [
        ("R_atm", _FP), ("T_down", _FP), ("T_up", _FP), ("s_alb", _FP),
        ("T_down_dir", _FP), ("R_atmQ", _FP), ("R_atmU", _FP),
    ]

def _fp(a):
    return np.ascontiguousarray(a, dtype=np.float32).ctypes.data_as(_FP)

def compute_lut(wl, aod, h2o, sza=30., vza=5., raa=90., altitude_km=1000.,
                atmo_model=1, aerosol_model=1, ozone_du=300.,
                surface_pressure=0., enable_polar=0):
    wl_  = np.asarray(wl,  dtype=np.float32)
    aod_ = np.asarray(aod, dtype=np.float32)
    h2o_ = np.asarray(h2o, dtype=np.float32)
    N    = wl_.size * aod_.size * h2o_.size
    Ra, Td, Tu, ss = [np.empty(N, np.float32) for _ in range(4)]
    cfg_c = _Cfg(
        wl=_fp(wl_), n_wl=wl_.size, aod=_fp(aod_), n_aod=aod_.size,
        h2o=_fp(h2o_), n_h2o=h2o_.size,
        sza=sza, vza=vza, raa=raa, altitude_km=altitude_km,
        atmo_model=atmo_model, aerosol_model=aerosol_model,
        ozone_du=ozone_du, surface_pressure=surface_pressure,
        enable_polar=enable_polar,
    )
    arr_c = _Arr(R_atm=_fp(Ra), T_down=_fp(Td), T_up=_fp(Tu), s_alb=_fp(ss))
    _lib.atcorr_compute_lut(ctypes.byref(cfg_c), ctypes.byref(arr_c))
    sh = (aod_.size, h2o_.size, wl_.size)
    return dict(cfg_c=cfg_c, arr_c=arr_c,
                wl=wl_, aod=aod_, h2o=h2o_,
                R_atm=Ra.reshape(sh), T_down=Td.reshape(sh),
                T_up=Tu.reshape(sh),  s_alb=ss.reshape(sh),
                _Ra=Ra, _Td=Td, _Tu=Tu, _ss=ss)

def lut_slice(lut, aod_val, h2o_val):
    n = lut['wl'].size
    Rs, Tds, Tus, ss = [np.empty(n, np.float32) for _ in range(4)]
    _lib.atcorr_lut_slice(
        ctypes.byref(lut['cfg_c']), ctypes.byref(lut['arr_c']),
        ctypes.c_float(aod_val), ctypes.c_float(h2o_val),
        _fp(Rs), _fp(Tds), _fp(Tus), _fp(ss), None)
    return dict(R_atm=Rs, T_down=Tds, T_up=Tus, s_alb=ss)

nrows, ncols = 64, 64
npix = nrows * ncols

# ── Synthetic AOD map with NaN outliers ───────────────────────────────────────
r_idx = np.repeat(np.arange(nrows), ncols).astype(np.float32)
c_idx = np.tile(np.arange(ncols), nrows).astype(np.float32)
aod_map = (0.10 + 0.15 * np.sin(r_idx * 0.3) * np.cos(c_idx * 0.3)).astype(np.float32)
aod_map[::100] = float('nan')   # cloud-masked outliers

# ── 1. Gaussian filter (in-place, NaN excluded from neighbourhood) ────────────
_lib.spatial_gaussian_filter.argtypes = [
    _FP, ctypes.c_int, ctypes.c_int, ctypes.c_float]
_lib.spatial_gaussian_filter.restype = None

aod_gauss = aod_map.copy()
n_nan_before = int(np.isnan(aod_gauss).sum())
_lib.spatial_gaussian_filter(
    _fp(aod_gauss), ctypes.c_int(nrows), ctypes.c_int(ncols), ctypes.c_float(2.5))
n_nan_after = int(np.isnan(aod_gauss).sum())
print(f"Gaussian filter: NaN before={n_nan_before}  after={n_nan_after}")

# ── 2. Box filter (input → separate output buffer) ────────────────────────────
_lib.spatial_box_filter.argtypes = [
    _FP, _FP, ctypes.c_int, ctypes.c_int, ctypes.c_int]
_lib.spatial_box_filter.restype = None

aod_box = np.empty(npix, np.float32)
_lib.spatial_box_filter(
    _fp(aod_map), _fp(aod_box),
    ctypes.c_int(nrows), ctypes.c_int(ncols), ctypes.c_int(3))
print(f"Box filter (half=3): output mean = {np.nanmean(aod_box):.4f}")

# ── 3. Terrain illumination correction ───────────────────────────────────────
_lib.cos_incidence.argtypes   = [ctypes.c_float] * 4
_lib.cos_incidence.restype    = ctypes.c_float
_lib.skyview_factor.argtypes  = [ctypes.c_float]
_lib.skyview_factor.restype   = ctypes.c_float
_lib.atcorr_terrain_T_down.argtypes = [ctypes.c_float] * 5
_lib.atcorr_terrain_T_down.restype  = ctypes.c_float
_lib.atcorr_terrain_T_up.argtypes   = [ctypes.c_float] * 3
_lib.atcorr_terrain_T_up.restype    = ctypes.c_float

sza, saa, slope, aspect = 35.0, 150.0, 20.0, 180.0   # south-facing slope
cos_i = _lib.cos_incidence(ctypes.c_float(sza), ctypes.c_float(saa),
                            ctypes.c_float(slope), ctypes.c_float(aspect))
V_d   = _lib.skyview_factor(ctypes.c_float(slope))
print(f"\nTerrain:  cos_i = {cos_i:.4f}  skyview = {V_d:.4f}"
      + ("  (topographic shadow)" if cos_i <= 0 else ""))

T_down, T_down_dir = 0.85, 0.70
cos_sza = math.cos(math.radians(sza))
T_down_eff = _lib.atcorr_terrain_T_down(
    ctypes.c_float(T_down), ctypes.c_float(T_down_dir),
    ctypes.c_float(cos_sza), ctypes.c_float(cos_i), ctypes.c_float(V_d))
print(f"T_down flat={T_down:.4f}  T_down_eff (slope)={T_down_eff:.4f}")

T_up = 0.90
T_up_eff = _lib.atcorr_terrain_T_up(
    ctypes.c_float(T_up), ctypes.c_float(math.cos(math.radians(8.0))),
    ctypes.c_float(15.0))
print(f"T_up flat={T_up:.4f}   T_up_eff (off-nadir view)={T_up_eff:.4f}")

# ── 4. Adjacency effect correction ───────────────────────────────────────────
_lib.adjacency_correct_band.argtypes = [
    _FP, ctypes.c_int, ctypes.c_int,
    ctypes.c_float, ctypes.c_float, ctypes.c_float, ctypes.c_float,
    ctypes.c_float, ctypes.c_float, ctypes.c_float,
    ctypes.c_float, ctypes.c_float,
]
_lib.adjacency_correct_band.restype = None

r_boa = (0.10 + 0.25 * c_idx / ncols).astype(np.float32).copy()
_lib.adjacency_correct_band(
    _fp(r_boa), ctypes.c_int(nrows), ctypes.c_int(ncols),
    ctypes.c_float(1.0),    # psf_radius_km
    ctypes.c_float(30.0),   # pixel_size_m (Landsat 30 m)
    ctypes.c_float(1.0 - T_down_dir),  # T_scat
    ctypes.c_float(0.10),   # s_alb
    ctypes.c_float(0.65),   # wl_um
    ctypes.c_float(0.15),   # aod550
    ctypes.c_float(1013.25),# pressure [hPa]
    ctypes.c_float(sza), ctypes.c_float(8.0))
print(f"\nAdjacency correction: mean BOA = {r_boa.mean():.4f}")

# ── 5. Uncertainty propagation ────────────────────────────────────────────────
lut = compute_lut(
    wl=[0.65], aod=[0.0, 0.1, 0.2, 0.4], h2o=[1.0, 3.0],
    sza=35., vza=8., raa=90., altitude_km=1000.,
    atmo_model=1, aerosol_model=1, ozone_du=300.,
)

_lib.sixs_E0.argtypes              = [ctypes.c_float]
_lib.sixs_E0.restype               = ctypes.c_float
_lib.sixs_earth_sun_dist2.argtypes = [ctypes.c_int]
_lib.sixs_earth_sun_dist2.restype  = ctypes.c_double
E0 = float(_lib.sixs_E0(ctypes.c_float(0.65)))
d2 = float(_lib.sixs_earth_sun_dist2(ctypes.c_int(185)))

_lib.uncertainty_compute_band.argtypes = [
    _FP, _FP,                         # rad_band (or NULL), refl_band
    ctypes.c_int,                     # npix
    ctypes.c_float, ctypes.c_float,   # E0, d2
    ctypes.c_float,                   # cos_sza
    ctypes.c_float, ctypes.c_float,   # T_down, T_up
    ctypes.c_float, ctypes.c_float,   # s_alb, R_atm
    ctypes.c_float, ctypes.c_float,   # nedl, aod_sigma
    ctypes.POINTER(_Cfg),
    ctypes.POINTER(_Arr),
    ctypes.c_float, ctypes.c_float, ctypes.c_float,  # wl_um, aod_val, h2o_val
    _FP,                              # sigma_out
]
_lib.uncertainty_compute_band.restype = None

sigma = np.empty(npix, np.float32)
_lib.uncertainty_compute_band(
    None, _fp(r_boa),
    ctypes.c_int(npix),
    ctypes.c_float(E0), ctypes.c_float(d2),
    ctypes.c_float(math.cos(math.radians(35.))),
    ctypes.c_float(0.85), ctypes.c_float(0.90),
    ctypes.c_float(0.08), ctypes.c_float(0.05),
    ctypes.c_float(0.), ctypes.c_float(0.04),
    ctypes.byref(lut['cfg_c']), ctypes.byref(lut['arr_c']),
    ctypes.c_float(0.65), ctypes.c_float(0.15), ctypes.c_float(2.0),
    _fp(sigma),
)
print(f"Uncertainty (σ_rfl) at 650 nm: mean = {np.nanmean(sigma):.5f}")
